rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,427 @@
+an explicit default value to operate will cause problems for older agents which will not know about the value; this is
+particularly true for Roles when adding a new field which determines access to a resource.
+
+If a client attempts to update a resource with the new field during a rolling deployment when both the old and new
+versions of Auth are running simultaneously there is a 50/50 chance that writes and reads which contain the new field
+will be lossy since the old instance doesn't know about the new field. However, data loss is a problem outside of
+rolling Auth updates. Reads are going to be lossy unless the client version is at the same version or newer than Auth
+because they won't know about new optional fields. Writes are going to be lossy unless Auth is at the same or newer
+version than clients because it won't know that there are additional fields to persist. Until there is a solution for
+data loss under normal operation of a cluster we should accept the possible data loss during the very small time frame
+which a rolling deployment occurs.
+
+The risks of data loss from new fields may be mitigated in the future by the Cloud First Deployment strategy defined in
+RFD 134 and automatic client updates. If Cloud is running off of master the probability of a client attempting to
+read/write a field that Auth doesn't know about is less likely. Ensuring that clients are using the appropriate version
+for a cluster will prevent client drift that results in data loss. As long as the clients are not upgraded until after
+Auth then there is no opportunity for new fields to be accessed.
+
+### Changing the meaning of a field
+
+Altering or extending the capabilities of a field will result in backward incompatible behavior. Imagine the following
+resource:
+
+```protobuf
+message Foo {
+  repeated Bar Bars = 1;
+}
+
+message Bar {
+  string Kind = 1;
+  string Namespace = 2;
+  string Name = 3;
+}
+```
+
+Adding a new kind of `Bar` in a newer version will force the old Auth instance to make a decision about how to handle
+unknown Bar kind. This is particularly troublesome when the resource is involved in access decisions. The only
+reasonable decision the old Auth server can make in this case is to deny access to prevent any possible bypasses as a
+result of not knowing how to process the new Bar.
+
+This is also true for fields which are just scalar types, adding a new value to a field that is a string will result
+in older instances being unable to parse the field. For example, there are several resources which take a predicate
+expression defined in a string; if we were to extend the supported expressions by adding new function(s) it would
+result in older instances being unable to parse the predicate expression and possibly prevent access to a resource
+that users should be granted access to.
+
+### Alternate Options Considered
```

Comment:
very minor nit: can we move this section to the end of the document?  When quickly scanning through it's difficult to tell at a glance wether a given section is an alternate option or part of the primary proposal.